Installation: improve FreeBSD instructions for Manual Build

jbeich opened the pull request.

jbeich committed:
Installation: unbreak Meson steps

- ninja doesn't support `--tags` unlike meson
- `--skip-subprojects` is more clear than `--tags runtime,man`

jbeich committed:
Installation: invoke Git when using Meson like with CMake

Trying to build in a random directory like HOME may fail or build
something else. So, provide the context for a user who skipped
reading CMake section.

jbeich commented:
Some meta notes:
- I don't like CMake, so don't expect me to fix Hyprland (unlike other hyprwm projects)
- Clang/libc++ 16 in FreeBSD **base** (unlike ports) is [currently WIP](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=271047) and may take ~1 year to propagate into any -RELEASE.

vaxerski commented:
huh why is cmake linux only?

jbeich commented:
> huh why is cmake linux only?

See https://github.com/hyprwm/Hyprland/pull/2209

fufexan commented:
@jbeich has anything changed since 4 months ago?

jbeich commented:
Not much:
- Both CMake and Meson instructions still work
- FreeBSD 13.3/14.0 haven't been released yet, so building with GCC is still easier

105b55239eba added `meson subprojects update` but it isn't very useful compared to explicit `git clone --recursive` (or `git submodule update --init --recursive`):
```
$ meson subprojects update --reset
Updating udis86...
  -> Cannot update subproject with no wrap file
Updating tracy...
  -> Cannot update subproject with no wrap file
Updating hyprland-protocols...
  -> Cannot update subproject with no wrap file
Initialized empty Git repository in /tmp/Hyprland/subprojects/wlroots/.git/
remote: Enumerating objects: 480, done.
remote: Counting objects: 100% (480/480), done.
remote: Compressing objects: 100% (466/466), done.
remote: Total 480 (delta 27), reused 160 (delta 5), pack-reused 0
Receiving objects: 100% (480/480), 773.01 KiB | 300.00 KiB/s, done.
Resolving deltas: 100% (27/27), done.
From https://gitlab.freedesktop.org/wlroots/wlroots
 * branch            c74f89d4f84bfed0284d3908aee5d207698c70c5 -> FETCH_HEAD
HEAD is now at c74f89d Avoid using memcpy() to copy structs
Applying diff file "wlroots-meson-build.patch"
Updating wlroots...
  -> c74f89d (grafted, HEAD) Avoid using memcpy() to copy structs [Simon Ser]

$ meson setup build
$ meson compile -C build
[...]
../subprojects/wlroots/include/wlr/render/egl.h:19:10: fatal error: 'wlr/config.h' file not found
#include <wlr/config.h>
         ^~~~~~~~~~~~~~

$ find . -name config.h
./build/subprojects/wlroots/include/wlroots/config.h
./build/subprojects/wlroots/include/config.h

$ cat subprojects/packagefiles/wlroots-meson-build.patch
[...]
-install_subdir('wlr',
+install_subdir('wlroots',
[...]
```

jbeich commented:
> Both CMake and Meson instructions still work

From this PR, otherwise Meson fails.

fufexan commented:
> but it's broken outside of Nix

Can I fix it somehow or is it better to just not use `meson subprojects`?

jbeich commented:
Better remove `wlroots.wrap`. It's not consistent with other subprojects (e.g., hyprland-protocols, udis86) or CMake support.

Downstream I use `git submodule` instead of `meson subprojects` and system wlroots (a la Nix), so don't need synthetic `soversion` bump.

fufexan commented:
Alright, I'll remove it soon (unless I forget).

jbeich committed:
Installation: deduplicate FreeBSD dependencies

https://github.com/freebsd/freebsd-ports/commit/1b744b687e87

jbeich committed:
Installation: add tomlplusplus on FreeBSD

https://github.com/hyprwm/Hyprland/commit/d36055054651
https://github.com/freebsd/freebsd-ports/commit/57bb94121dc2

jbeich committed:
Installation: temporarily pin GCC version on FreeBSD

Until GCC_DEFAULT is bumped to 14 or later in
https://github.com/freebsd/freebsd-ports/blob/main/Mk/bsd.default-versions.mk

jbeich committed:
Installation: drop GCC dependency on FreeBSD

System tomlplusplus is built against libc++ which conflicts with
libstdc++ used by GCC. While tomlplusplus can be bundled it's only
easy to do via Meson (--force-fallback-for).

/usr/local/bin/ld: hyprpm/CMakeFiles/hyprpm.dir/src/core/DataState.cpp.o: in function `DataState::addNewPluginRepo(SPluginRepository const&)':
DataState.cpp:(.text+0x16b4): undefined reference to `toml::v3::table::get_lower_bound(std::basic_string_view<char, std::char_traits<char> >)'
/usr/local/bin/ld: DataState.cpp:(.text+0x1793): undefined reference to `toml::v3::table::insert_with_hint(toml::v3::impl::table_iterator<true>, toml::v3::key&&, std::unique_ptr<toml::v3::node, std::default_delete<toml::v3::node> >&&)'
/usr/local/bin/ld: DataState.cpp:(.text+0x2c59): undefined reference to `toml::v3::impl::print_to_stream(std::ostream&, toml::v3::table const&)'
/usr/local/bin/ld: hyprpm/CMakeFiles/hyprpm.dir/src/core/DataState.cpp.o: in function `DataState::pluginRepoExists(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
DataState.cpp:(.text+0x3739): undefined reference to `toml::v3::ex::parse_file(std::basic_string_view<char, std::char_traits<char> >)'
/usr/local/bin/ld: DataState.cpp:(.text+0x3783): undefined reference to `toml::v3::table::get(std::basic_string_view<char, std::char_traits<char> >)'
/usr/local/bin/ld: DataState.cpp:(.text+0x37ab): undefined reference to `toml::v3::table::get(std::basic_string_view<char, std::char_traits<char> >)'
[...]

jbeich commented:
Rebased and adjusted dependencies to the current Hyprland and FreeBSD versions.

fufexan commented:
Is there anything left to do in this PR or is it ready? @jbeich

jbeich commented:
Tested both Meson and CMake instructions - still good.

jbeich commented:
Wait, `--skip-subprojects` may be wrong for the plugin API.

fufexan commented:
Is there anything that can be done on the meson side, so we don't have to include all the commands here? The wlroots.wrap file used to patch the subproject so it properly installed headers, but now I don't know of a better approach.

jbeich committed:
Installation: add hypr* dependencies on FreeBSD

https://github.com/hyprwm/Hyprland/commit/13f6f0b923ff
https://github.com/hyprwm/Hyprland/commit/c3882bb83240
https://github.com/hyprwm/Hyprland/commit/ea954494024c
https://github.com/hyprwm/Hyprland/commit/8c64a4bad710

jbeich committed:
Installation: drop wlroots leftovers on FreeBSD

https://github.com/hyprwm/Hyprland/commit/016da234d0e8

jbeich committed:
Installation: add libuuid on FreeBSD

FreeBSD uuid(3) has slightly different API than Linux and doesn't support
random UUIDs unlike NetBSD. Instead use external implementation.

https://github.com/hyprwm/Hyprland/commit/7778f01194d8

jbeich committed:
Installation: add xcb-icccm and xcb-errors on FreeBSD

jbeich committed:
Installation: drop GNU make on FreeBSD

https://github.com/hyprwm/Hyprland/commit/c03dc47124ca

jbeich commented:
I've dropped Meson workaround and rebased after aquamarine switch.

fufexan commented on pages/Getting Started/Installation.md:
```diff
@@ -370,10 +370,11 @@ installed._
-ninja -C build
-ninja -C build install --tags runtime,man
+meson compile -C build
+meson install -C build --skip-subprojects
```
Does this not mean udis86 will not be found?

jbeich replied:
udis86 is not part of Hyprland plugin API unlike wlroots was before aquamarine. `--skip-subprojects` is still necessary to avoid installing hyprland-protocols (plugins use preprocessed headers instead). Hyprland itself doesn't support building against system-wide copy of udis86 ([Alpine patches this](https://gitlab.alpinelinux.org/alpine/aports/-/blob/master/testing/hyprland/use-system-installed-udis86.patch)), and the bundled version uses CMake to build the submodule.

fufexan replied:
That's fine by me. However, I've made progress towards allowing Hyprland to build with a system-wide copy of udis86 (provided it's canihavesomecoffee's fork, which includes a .pc file). See this PR for more info about it, maybe leave your thoughts as well https://github.com/hyprwm/Hyprland/pull/7885.
